packages/playwright-core/src/server/injected/recorder.ts

```diff
@@ -442,217 +442,181 @@ class RecordActionTool implements RecorderTool {
   }
 
-  onDragStart(event: DragEvent) {
-    consumeEvent(event);
-  }
-
   onKeyDown(event: KeyboardEvent) {
-    if (event.key === 'Escape') {
-      const selection = this._recorder.document.getSelection();
-      if (selection && selection.toString())
-        this._resetSelectionAndHighlight();
-      else
-        this._recorder.delegate.setMode?.('recording');
-      consumeEvent(event);
-      return;
-    }
-
-    if (event.key === 'Enter') {
-      this._commitAction();
-      consumeEvent(event);
-      return;
-    }
-
-    // Only allow keys that control text selection.
-    if (!['ArrowLeft', 'ArrowUp', 'ArrowRight', 'ArrowDown', 'Shift', 'Control', 'Meta', 'Alt', 'AltGraph'].includes(event.key)) {
-      consumeEvent(event);
-      return;
-    }
-  }
-
-  onKeyUp(event: KeyboardEvent) {
     consumeEvent(event);
   }
 
-  onScroll(event: Event) {
-    this._hoverHighlight = null;
-    this._showHighlight(false);
-  }
-
-  private _generateAction(): actions.Action | null {
-    if (this._inputHighlight) {
-      const target = this._inputHighlight.elements[0] as HTMLInputElement;
-      if (target.nodeName === 'INPUT' && ['checkbox', 'radio'].includes(target.type.toLowerCase())) {
+  private _generateAction(): actions.AssertAction | null {
+    const target = this._hoverHighlight?.elements[0];
+    if (!target)
+      return null;
+    if (target.nodeName === 'INPUT' || target.nodeName === 'TEXTAREA') {
+      const { selector } = generateSelector(this._recorder.injectedScript, target, { testIdAttributeName: this._recorder.state.testIdAttributeName });
+      if (target.nodeName === 'INPUT' && ['checkbox', 'radio'].includes((target as HTMLInputElement).type.toLowerCase())) {
         return {
           name: 'assertChecked',
-          selector: this._inputHighlight.selector,
+          selector,
           signals: [],
           // Interestingly, inputElement.checked is reversed inside this event handler.
-          checked: !(target as HTMLInputElement).checked,
+          checked: (target as HTMLInputElement).checked,
         };
       } else {
         return {
           name: 'assertValue',
-          selector: this._inputHighlight.selector,
+          selector,
           signals: [],
-          value: target.value,
+          value: (target as HTMLInputElement).value,
         };
       }
-    } else if (this._selectionText && this._selectionHighlight) {
+    } else {
+      const { selector } = generateSelector(this._recorder.injectedScript, target, { testIdAttributeName: this._recorder.state.testIdAttributeName, forTextExpect: true });
       return {
         name: 'assertText',
-        selector: this._selectionHighlight.selector,
+        selector,
         signals: [],
-        text: this._selectionText.selectedText,
-        substring: this._selectionText.fullText !== this._selectionText.selectedText,
+        text: target.textContent!,
```

Comment:
To make sure we match our own definition of element text:

```suggestion
        text: elementText(new Map(), target).full,
```